### PR TITLE
SAK-45616 add Proctorio capabilities to Sakai

### PR DESCRIPTION
**Pull request by ottenhoff**
Please preserve the two commits via rebase + merge

### CODE REVIEW

**Comment by bjones86** [comment]
@ottenhoff what's the difference between this and #9400?

**Comment by ottenhoff** [comment]
This PR has the @bbailla2 commit 

Merger: please preserve the two commits to preserve the two authors

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java [hunk 1]
```diff
@@ -348,6 +401,18 @@ private void handlePlugin( String secureDeliveryPlugin ) {
 	
 		try
 		{
+			// This is a built-in integration with no additional JAR file
+			if (secureDeliveryPlugin.equalsIgnoreCase("proctorio")) {
+				ApplicationContext appCtx = new AnnotationConfigApplicationContext(SecureDeliveryProctorio.class);
```
Comment (ern) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:406]: samigo-services is a lib in shared and creating an application context in this way will have consequences.
For example components are managed by ComponentManager and webapps have their own init tied to the servlet lifecycle. ApplicationContexts are managed by these entities meaning they have a start and they have an end.
This application context is continually generated with every call here.

**Comment by bjones86** [comment]
I'm unsure of whether to approve this, or if it needs changes. To date, we have not been able to test this properly at Western. We're still waiting to hear back from Proctorio about certain things. The last I heard (from Proctorio) was that the code was not generating the student URLs correctly, and that it likely warranted code changes. However I don't know if this is accurate, or if the changes have already been made. They indicated they would contact Longsight about this.

We were also told about some sakai.properties for cookies and X-frame options. However, I'm unsure if we've tested these, or if we even want to set these.

There is still also the outstanding concern that Earle brought up. I understand Sam tried to address this, but was unable to after many hours. Is this something Earle could help with?

**Comment by ottenhoff** [comment]
* Rebased with latest
* Remove caching of links
* Removed persistence of links

**Comment by ottenhoff** [comment]
> We were also told about some sakai.properties for cookies and X-frame options. However, I'm unsure if we've tested these, or if we even want to set these.

X-Frame-Options: if you want to allow a service or portal or some site to frame your LMS, then you need to alter X-Frame-Options.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java (hunk 1 above)
Comment (ern) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/shared/impl/assessment/SecureDeliveryServiceImpl.java:406]: Resolving this comment as this was the pattern established here long before this PR. In Spring 5 there is [method](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/context/support/GenericApplicationContext.html#registerBean-java.lang.Class-org.springframework.beans.factory.config.BeanDefinitionCustomizer...-) for registering beans with the application context. 

I think the large questions to ask here is if this design is mean to look for and add a service to an existing spring context (either Sakai Application Context or a Web Application Context) or is it simply wanting to inject Sakai services that are needed (use SpringBeanAutowiringSupport).

**Comment by ottenhoff** [comment]
@bjones86 I'd like to see this get into trunk so it can be maintained in trunk going forward instead of in custom branches. We're committed to fixing bugs and Earle has removed his objection to the old wiring scheme.

**Review by ern — APPROVED** [review]

**Review by bjones86 — APPROVED** [review]

**Comment by ern** [comment]
@ottenhoff can you squash your commits into 1 so that there is a single commit from each author?

**Comment by ottenhoff** [comment]
I tried but failed to squash into two commits. Can you just make sure to give a shout out to @bbailla2 in the final commit message?

**Pull request merged**